File: python_prototype/solution_postprocess.py
from game import Game, GameOperation, OperationStepForward, GameNode, GameNodeType
from solver import SearchState
from typing import TypeGuard, List, Tuple
import networkx as nx

import matplotlib.pyplot as plt
from networkx.drawing.nx_pydot import graphviz_layout

def is_ops_all_step_forward(xs: list[GameOperation]) -> TypeGuard[list[OperationStepForward]]:
    return all(isinstance(x, OperationStepForward) for x in xs)

def solution_to_graph(input_solution: SearchState) -> nx.DiGraph:
    assert is_ops_all_step_forward(input_solution.path)

    # Reconstruct intermediate game states in order from start to end
    states = []
    game_ptr = input_solution.state_game
    while game_ptr is not None:
        states.append(game_ptr)
        game_ptr = game_ptr.previous_state
    states.reverse()
    assert len(states) == len(input_solution.path) + 1

    G = nx.DiGraph()
    for i, op in enumerate(input_solution.path):
        src_group = states[i].groups[op.src]
        op_item = src_group[-1]
        next_src_group = states[i + 1].groups[op.src]
        op_revealing_color = (
            next_src_group[-1].color if len(next_src_group) > 0 else None
        )
        G.add_node(
            i,
            label=f"{op}",
            op_src=op.src,
            op_dst=op.dst,
            op_color=op_item.color,
            op_revealing_color=op_revealing_color,
        )

    G.add_node("s", label="s")
    G.add_node("t", label="t")

    last_op_on_group_index = [-1] * len(input_solution.state_game.groups)

    for i, op in enumerate(input_solution.path):
        if (es_1 := last_op_on_group_index[op.src]) != -1:
            G.add_edge(es_1, i)
        if (es_2 := last_op_on_group_index[op.dst]) != -1:
            G.add_edge(es_2, i)
        if es_1 == es_2 == -1:
            G.add_edge("s", i)
        last_op_on_group_index[op.src] = last_op_on_group_index[op.dst] = i

    src_nodes = set()
    for src, _ in G.edges:
        src_nodes.add(src)

    for i in range(len(input_solution.path)):
        if i not in src_nodes:
            G.add_edge(i, "t")

    TR = nx.transitive_reduction(G)
    TR.add_nodes_from(G.nodes(data=True))

    return TR

def solution_postprocess(input_solution: SearchState, initial_game: Game) -> nx.DiGraph:
    G = solution_to_graph(input_solution)

    # Produce the prioritized topological order once
    ordered_nodes: list[int] = list(priority_topo_sort(G, initial_game))

    # Prefix labels with sequence numbers for readability
    for ind, node in enumerate(ordered_nodes, start=1):
        node_obj = G.nodes[node]
        node_obj['seq_index'] = ind
        node_obj['label'] = f"{ind:02d}: {node_obj['label']}"

    # Compute collapsable flags using a simulated game state
    game_state = initial_game
    for node in ordered_nodes:
        node_data = G.nodes[node]

        forward_ops = [
            op for op in game_state.ops() if isinstance(op, OperationStepForward)
        ]
        if len(forward_ops) == 1:
            only = forward_ops[0]
            collapsable = (only.src == node_data.get('op_src') and only.dst == node_data.get('op_dst'))
        else:
            collapsable = False
        node_data['collapsable'] = collapsable

        # Advance simulation
        game_state = game_state.apply_op(
            OperationStepForward(node_data.get('op_src'), node_data.get('op_dst'))
        )

    return G


def build_solution_summaries(input_solution: SearchState, initial_game: Game) -> List[Tuple[str, list[OperationStepForward]]]:
    """Return grouped English summaries and their operations.

    - Prefers grouping consecutive steps that merge the same color into the
      same destination tube: "Merge #RRGGBB from tubes S... into tube D".
    - Otherwise groups consecutive steps that empty the same source tube:
      "Empty tube S into tubes D...".
    - Appends "(completes tube)" if any step in the group completes a tube.
    - Returns a list of (summary, [OperationStepForward...]) following the
      prioritized step order.
    """
    G = solution_to_graph(input_solution)
    ordered_nodes: list[int] = list(priority_topo_sort(G, initial_game))

    def color_to_hex(c: tuple[int, int, int] | None) -> str:
        if isinstance(c, tuple):
            return f"#{c[0]:02X}{c[1]:02X}{c[2]:02X}"
        return "unknown color"

    # Build linear step info with post-move completion flags
    steps: list[dict] = []
    sim = initial_game
    for node in ordered_nodes:
        nd = G.nodes[node]
        src = nd.get('op_src')
        dst = nd.get('op_dst')
        color = nd.get('op_color') if isinstance(nd.get('op_color'), tuple) else None
        op = OperationStepForward(src, dst)
        sim_next = sim.apply_op(op)
        completes = sim_next.is_group_completed(sim_next.groups[dst])
        steps.append({
            'src': src,
            'dst': dst,
            'color': color,
            'color_hex': color_to_hex(color),
            'op': op,
            'completes': completes,
        })
        sim = sim_next

    n = len(steps)
    i = 0
    results: List[Tuple[str, list[OperationStepForward]]] = []

    while i < n:
        # Try to extend two possible run types from i
        c0 = steps[i]['color']
        d0 = steps[i]['dst']
        s0 = steps[i]['src']

        # Length if grouping by (color, dst)
        j = i + 1
        while j < n and steps[j]['color'] == c0 and steps[j]['dst'] == d0:
            j += 1
        merge_len = j - i

        # Length if grouping by (src)
        k = i + 1
        while k < n and steps[k]['src'] == s0:
            k += 1
        empty_len = k - i

        mode = 'merge' if merge_len >= empty_len else 'empty'
        end = (i + merge_len) if mode == 'merge' else (i + empty_len)
        segment = steps[i:end]

        if mode == 'merge':
            # Merge #color from sources into a single destination
            color_hex = segment[0]['color_hex']
            dst = segment[0]['dst'] + 1
            sources = sorted({s['src'] + 1 for s in segment})
            sources_str = ", ".join(map(str, sources))
            summary = f"Merge {color_hex} from tubes {sources_str} into tube {dst}"
        else:
            # Empty source into multiple destinations
            src = segment[0]['src'] + 1
            dests = sorted({s['dst'] + 1 for s in segment})
            dests_str = ", ".join(map(str, dests))
            summary = f"Empty tube {src} into tubes {dests_str}"

        if any(s['completes'] for s in segment):
            summary += " (completes tube)"

        ops = [s['op'] for s in segment]
        results.append((summary, ops))
        i = end

    return results

def priority_topo_sort(G: nx.DiGraph, initial_game: Game):
    """Yield a prioritised topological ordering of ``G``.

    This function implements a Kahn style algorithm but, when multiple nodes
    have zero in-degree, it scores them using information from the previously
    selected operation and the current game state.  The candidate with the
    highest score is dequeued and yielded immediately, so the caller can
    consume the order lazily.

    The scoring priorities are, from high to low:

    1. The move completes a tube in the current game state.
    2. ``op_color`` equals the previous operation's ``op_color``.
    3. ``op_color`` equals the previous operation's ``op_revealing_color``.
    4. ``{op_src, op_dst}`` intersects with that of the previous operation.

    Nodes ``"s"`` and ``"t"`` are skipped in the yielded order.
    """

    indegree = {node: G.in_degree(node) for node in G.nodes}
    zero_indegree: list = [n for n, d in indegree.items() if d == 0]

    prev_color = None
    prev_reveal = None
    prev_groups: set | None = None
    game_state = initial_game

    def score(node) -> int:
        """Score a candidate node based on the priorities above."""

        if node in {"s", "t"}:
            return -1
        node_data = G.nodes[node]
        node_color = node_data.get("op_color")
        node_src = node_data.get("op_src")
        node_dst = node_data.get("op_dst")
        s = 0

        # Top priority: completing a tube after applying the move
        op = OperationStepForward(node_src, node_dst)
        if game_state.is_group_completed(game_state.apply_op(op).groups[node_dst]):
            s += 8
        if prev_color is not None and node_color == prev_color:
            s += 4
        if prev_reveal is not None and node_color == prev_reveal:
            s += 2
        if prev_groups and (node_src in prev_groups or node_dst in prev_groups):
            s += 1
        return s

    while zero_indegree:
        if "s" in zero_indegree:
            current = "s"
            zero_indegree.remove("s")
        else:
            current = max(
                zero_indegree,
                key=lambda n: (score(n), -n if isinstance(n, int) else float("-inf")),
            )
            zero_indegree.remove(current)

        for succ in G.successors(current):
            indegree[succ] -= 1
            if indegree[succ] == 0:
                zero_indegree.append(succ)

        if current in {"s", "t"}:
            continue

        node_data = G.nodes[current]
        prev_color = node_data.get("op_color")
        prev_reveal = node_data.get("op_revealing_color")
        prev_groups = {node_data.get("op_src"), node_data.get("op_dst")}

        # Advance the simulated game state to reflect the yielded operation
        game_state = game_state.apply_op(
            OperationStepForward(node_data.get("op_src"), node_data.get("op_dst"))
        )

        yield current

def show_graph(G: nx.DiGraph):
    pos = graphviz_layout(G, prog="dot", root=G.nodes["s"])
    node_colors = []
    edge_colors = []
    for n in G.nodes:
        if n in {"s", "t"}:
            node_colors.append("lightgray")
            edge_colors.append("black")
        else:
            color = G.nodes[n].get("op_color")
            if isinstance(color, tuple):
                node_colors.append([c / 255 for c in color])
            else:
                node_colors.append("lightgray")

            reveal_color = G.nodes[n].get("op_revealing_color")
            if isinstance(reveal_color, tuple):
                edge_colors.append([c / 255 for c in reveal_color])
            else:
                edge_colors.append("lightgray")
    nx.draw(
        G,
        pos,
        with_labels=True,
        arrows=True,
        node_color=node_colors,
        edgecolors=edge_colors,
        linewidths=2,
        labels=dict(G.nodes.data("label")),
    )
    plt.show()

File: src/ts/solution-postprocess.ts
// Solution postprocessing - TypeScript port of Python solution_postprocess.py

import { DirectedGraph } from './graph.ts';
import { Game, StepOp } from './game.ts';
import { SearchState } from './solver.ts';
import { Color } from './types.ts';

interface StepNodeData {
    label: string;
    opSrc: number;
    opDst: number;
    opColor: Color | null;
    opRevealingColor: Color | null;
    seqIndex?: number;
    collapsable?: boolean;
}

function isStepOp(op: StepOp | any): op is StepOp {
    return op instanceof StepOp;
}

export function solutionToGraph(inputSolution: SearchState): DirectedGraph<StepNodeData> {
    // Ensure all operations are StepOp (no undo operations)
    if (!inputSolution.path.every(isStepOp)) {
        throw new Error('Solution contains non-StepOp operations');
    }

    const stepPath = inputSolution.path as StepOp[];
    
    // Reconstruct intermediate game states in order from start to end
    const states: Game[] = [];
    let gamePtr: Game | null = inputSolution.stateGame;
    while (gamePtr !== null) {
        states.push(gamePtr);
        gamePtr = gamePtr.undoTargetState;
    }
    states.reverse();
    
    if (states.length !== stepPath.length + 1) {
        throw new Error(`State count mismatch: ${states.length} vs ${stepPath.length + 1}`);
    }

    const G = new DirectedGraph<StepNodeData>();
    
    // Create nodes for each operation
    for (let i = 0; i < stepPath.length; i++) {
        const op = stepPath[i];
        const srcGroup = states[i].groups[op.src];
        const opItem = srcGroup[srcGroup.length - 1];
        const nextSrcGroup = states[i + 1].groups[op.src];
        const opRevealingColor = nextSrcGroup.length > 0 ? nextSrcGroup[nextSrcGroup.length - 1].color : null;
        
        G.addNode(i.toString(), {
            label: `${op.toString()}`,
            opSrc: op.src,
            opDst: op.dst,
            opColor: opItem.color,
            opRevealingColor: opRevealingColor
        });
    }

    // Add special start and end nodes
    G.addNode("s", { label: "s", opSrc: -1, opDst: -1, opColor: null, opRevealingColor: null });
    G.addNode("t", { label: "t", opSrc: -1, opDst: -1, opColor: null, opRevealingColor: null });

    // Track last operation on each group
    const lastOpOnGroupIndex: number[] = new Array(inputSolution.stateGame.groups.length).fill(-1);

    // Build edges based on operation dependencies
    for (let i = 0; i < stepPath.length; i++) {
        const op = stepPath[i];
        const es1 = lastOpOnGroupIndex[op.src];
        const es2 = lastOpOnGroupIndex[op.dst];
        
        if (es1 !== -1) {
            G.addEdge(es1.toString(), i.toString());
        }
        if (es2 !== -1) {
            G.addEdge(es2.toString(), i.toString());
        }
        if (es1 === -1 && es2 === -1) {
            G.addEdge("s", i.toString());
        }
        
        lastOpOnGroupIndex[op.src] = i;
        lastOpOnGroupIndex[op.dst] = i;
    }

    // Connect operations with no successors to the end node
    const srcNodes = new Set<string>();
    for (const [from, to] of G.getAllEdges()) {
        srcNodes.add(from);
    }

    for (let i = 0; i < stepPath.length; i++) {
        if (!srcNodes.has(i.toString())) {
            G.addEdge(i.toString(), "t");
        }
    }

    // Apply transitive reduction
    const TR = G.transitiveReduction();
    
    return TR;
}

export function priorityTopoSort(G: DirectedGraph<StepNodeData>, initialGame: Game): string[] {
    let prevColor: Color | null = null;
    let prevReveal: Color | null = null;
    let prevGroups: Set<number> | null = null;
    let gameState = initialGame;

    const scoreFn = (nodeId: string): number => {
        if (nodeId === "s" || nodeId === "t") {
            return -1;
        }
        
        const nodeData = G.getNodeData(nodeId);
        if (!nodeData) return -1;
        
        const { opSrc, opDst, opColor } = nodeData;
        let score = 0;

        // Priority 1: completing a tube after applying the move (score +8)
        const op = new StepOp(opSrc, opDst);
        const nextGameState = gameState.apply(op);
        if (nextGameState.isGroupCompleted(nextGameState.groups[opDst])) {
            score += 8;
        }

        // Priority 2: same color as previous move (score +4)
        if (prevColor !== null && opColor !== null && opColor.toString() === prevColor.toString()) {
            score += 4;
        }

        // Priority 3: reveals expected color (score +2)
        if (prevReveal !== null && opColor !== null && opColor.toString() === prevReveal.toString()) {
            score += 2;
        }

        // Priority 4: uses same tubes as previous move (score +1)
        if (prevGroups && (prevGroups.has(opSrc) || prevGroups.has(opDst))) {
            score += 1;
        }

        return score;
    };

    const result: string[] = [];
    const inDegree = new Map<string, number>();
    
    // Initialize in-degrees
    for (const [nodeId] of G.getNodes()) {
        inDegree.set(nodeId, G.getInDegree(nodeId));
    }
    
    // Find nodes with zero in-degree
    const zeroInDegree: string[] = [];
    for (const [nodeId, degree] of inDegree) {
        if (degree === 0) {
            zeroInDegree.push(nodeId);
        }
    }

    while (zeroInDegree.length > 0) {
        let current: string;
        
        if (zeroInDegree.includes("s")) {
            current = "s";
            zeroInDegree.splice(zeroInDegree.indexOf("s"), 1);
        } else {
            // Sort by score (higher first), then by node id (lower first)
            zeroInDegree.sort((a, b) => {
                const scoreA = scoreFn(a);
                const scoreB = scoreFn(b);
                if (scoreA !== scoreB) return scoreB - scoreA;  // Higher score first
                return parseInt(a) - parseInt(b);  // Lower node id first
            });
            current = zeroInDegree.shift()!;
        }

        // Process successors
        for (const successor of G.getSuccessors(current)) {
            const newDegree = inDegree.get(successor)! - 1;
            inDegree.set(successor, newDegree);
            if (newDegree === 0) {
                zeroInDegree.push(successor);
            }
        }

        // Skip start and end nodes in output
        if (current === "s" || current === "t") {
            continue;
        }

        // Update state for next iteration
        const nodeData = G.getNodeData(current);
        if (nodeData) {
            prevColor = nodeData.opColor;
            prevReveal = nodeData.opRevealingColor;
            prevGroups = new Set([nodeData.opSrc, nodeData.opDst]);

            // Advance the simulated game state
            const op = new StepOp(nodeData.opSrc, nodeData.opDst);
            gameState = gameState.apply(op);
        }

        result.push(current);
    }

    return result;
}

export function solutionPostprocess(inputSolution: SearchState, initialGame: Game): DirectedGraph<StepNodeData> {
    const G = solutionToGraph(inputSolution);
    
    // Produce the prioritized topological order
    const orderedNodes: string[] = priorityTopoSort(G, initialGame);

    // Prefix labels with sequence numbers for readability
    for (let ind = 0; ind < orderedNodes.length; ind++) {
        const nodeId = orderedNodes[ind];
        const nodeData = G.getNodeData(nodeId);
        if (nodeData) {
            nodeData.seqIndex = ind + 1;
            nodeData.label = `${(ind + 1).toString().padStart(2, '0')}: ${nodeData.label}`;
        }
    }

    // Compute collapsable flags using a simulated game state
    let gameState = initialGame;
    for (const nodeId of orderedNodes) {
        const nodeData = G.getNodeData(nodeId);
        if (!nodeData) continue;

        const forwardOps = gameState.ops().filter(isStepOp);
        
        let collapsable = false;
        if (forwardOps.length === 1) {
            const only = forwardOps[0];
            collapsable = (only.src === nodeData.opSrc && only.dst === nodeData.opDst);
        }
        
        nodeData.collapsable = collapsable;

        // Advance simulation
        const op = new StepOp(nodeData.opSrc, nodeData.opDst);
        gameState = gameState.apply(op);
    }

    return G;
}